pkg/kafka/json_writer.go

```diff
@@ -35,28 +43,46 @@ type JSONWriter struct {
 	producer Producer
 }
 
-func newProducer(config *Config) Producer {
+func getMechanism(config *Config) (sasl.Mechanism, error) {
+	if config.ClientID == "" || config.ClientSecret == "" {
+		// no credentials set, possibly using unauthenticated connection
+		return nil, nil
+	}
+	if config.SaslMechanism == saslPlain {
+		return &plain.Mechanism{
+			Username: config.ClientID,
+			Password: config.ClientSecret,
+		}, nil
+	}
+	if config.SaslMechanism == saslScram {
+		return scram.Mechanism(scram.SHA512, config.ClientID, config.ClientSecret)
+
+	}
+	return nil, fmt.Errorf("sasl mechanism %s is not valid", config.SaslMechanism)
+}
 
+func newProducer(config *Config) (Producer, error) {
+	brokers := strings.Split(config.BootstrapServer, ",")
 	writer := &kafka.Writer{
-		Addr:         kafka.TCP(config.BootstrapServer),
+		Addr:         kafka.TCP(brokers...),
 		Topic:        config.Topic,
 		Balancer:     &kafka.ReferenceHash{},
 		Compression:  compress.Gzip,
 		Async:        true,
 		WriteTimeout: WriteTimeout,
 	}
-	if config.ClientID != "" && config.ClientSecret != "" {
-		mechanism := &plain.Mechanism{
-			Username: config.ClientID,
-			Password: config.ClientSecret,
-		}
+	mechanism, err := getMechanism(config)
+	if err != nil {
+		return writer, err
```

Comment:
I'd say that this should return `nil, err`. In general if a function returns an error it shouldn't also return the partially created result.